pkg/helper/cache.go

```diff
@@ -0,0 +1,226 @@
+	if len(gvk.Kind) == 0 {
+		return schema.GroupKind{}, "", "", "", fmt.Errorf("unable to determine GroupKind of %T", obj)
+	}
+
+	return gvk.GroupKind(), metadata.GetName(), metadata.GetNamespace(), resourceHash, nil
+}
+
+func getResourceVersionGeneration(obj runtime.Object) (string, int64, error) {
+	if obj == nil {
+		return "", 0, fmt.Errorf("nil object has no resourceVersion")
+	}
+	metadata, err := meta.Accessor(obj)
+	if err != nil {
+		return "", 0, err
+	}
+	if metadata == nil || reflect.ValueOf(metadata).IsNil() {
+		return "", 0, fmt.Errorf("object has no metadata")
+	}
+	rv := metadata.GetResourceVersion()
+	if len(rv) == 0 {
+		return "", 0, fmt.Errorf("missing resourceVersion")
+	}
+
+	return rv, metadata.GetGeneration(), nil
+}
+
+func (c *WorkResourceCache) UpdateCachedResourceMetadata(required runtime.Object, actual runtime.Object) {
+	if c == nil || c.cache == nil {
+		return
+	}
+	if required == nil || actual == nil {
+		return
+	}
+
+	c.rwLock.Lock()
+	defer c.rwLock.Unlock()
+
+	kind, name, namespace, resourceHash, err := getResourceMetadata(required)
+	if err != nil {
+		return
+	}
+	cacheKey := cachedVersionKey{
+		name:      name,
+		namespace: namespace,
+		kind:      kind,
+	}
+
+	resourceVersion, generation, err := getResourceVersionGeneration(actual)
+	if err != nil {
+		klog.V(4).Infof("error reading resourceVersion %s:%s:%s %s", name, kind, namespace, err)
+		return
+	}
+
+	c.cache[cacheKey] = cachedResource{resourceHash, resourceVersion, generation}
+	klog.V(7).Infof("updated resourceVersion of %s:%s:%s %s", name, kind, namespace, resourceVersion)
+}
+
+// in the circumstance that an ApplyFoo's 'required' is the same one which was previously
+// applied for a given (name, kind, namespace) and the existing resource (if any),
+// hasn't been modified since the ApplyFoo last updated that resource, then return true (we don't
+// need to reapply the resource). Otherwise return false.
+func (c *WorkResourceCache) SafeToSkipApply(required runtime.Object, existing runtime.Object) bool {
+	if c == nil || c.cache == nil {
+		return false
+	}
+	if required == nil || existing == nil {
+		return false
+	}
+
+	c.rwLock.RLock()
+	defer c.rwLock.RUnlock()
+
+	kind, name, namespace, resourceHash, err := getResourceMetadata(required)
+	if err != nil {
+		return false
+	}
+	cacheKey := cachedVersionKey{
+		name:      name,
+		namespace: namespace,
+		kind:      kind,
+	}
+
+	resourceVersion, _, err := getResourceVersionGeneration(existing)
+	if err != nil {
+		return false
+	}
+
+	var versionMatch, hashMatch bool
+	if cached, exists := c.cache[cacheKey]; exists {
+		versionMatch = cached.resourceVersion == resourceVersion
+		hashMatch = cached.resourceHash == resourceHash
+		if versionMatch && hashMatch {
+			klog.V(4).Infof("found matching resourceVersion & manifest hash")
+			return true
+		}
+	}
+
+	return false
+}
+
+func (c *WorkResourceCache) SafeToSkipApplyWithGeneration(required runtime.Object, existing runtime.Object) bool {
```

Comment:
This needs godoc explaining that it is only safe for cases where
1. the generation is used
2. the generation is honored for the fields you're concerned about
3. how to figure out if generation does what you want